<launch>
  <arg name="viewer" default="false" /> <!-- Must be globally qualified -->
  <arg name="manager" default="ps4eye_manager_remote" /> <!-- Must be globally qualified -->
  <arg name="respawn" default="false" />
  <!-- TODO Arguments for debayer, interpolation methods? -->

  <arg     if="$(arg respawn)" name="bond" value="" />
  <arg unless="$(arg respawn)" name="bond" value="--no-bond" />

  <node pkg="nodelet" type="nodelet" name="$(arg manager)" args="manager" output="screen" />

  <!-- arguments for camera param uri -->
  <arg name="camera_info_file_default"
       default="$(find jsk_aero_startup)/config/ps4eye_default.yaml" />
  <arg name="camera_info_file_right"
       default="$(find jsk_aero_startup)/config/ps4eye_right.yaml" />
  <arg name="camera_info_file_left"
       default="$(find jsk_aero_startup)/config/ps4eye_left.yaml" />
  <arg name="camera_info_url_default" default="file://$(arg camera_info_file_default)" />
  <arg name="camera_info_url_right" default="file://$(arg camera_info_file_right)" />
  <arg name="camera_info_url_left" default="file://$(arg camera_info_file_left)" />

  <!-- arguments for tf coords -->
  <arg name="parent_frame" default="/head_base_link" />
  <arg name="parent_transform" value="0.06 0.04 0.08 -1.570791 0  -1.570791"/>

  <!-- This launchfile should bring up a node that broadcasts a ros image
       transport on /webcam/image_raw -->
  <arg name="load_driver" default="true" />
  <arg name="DEVICE" default="/dev/ps4eye"/>
  <!-- The GStreamer framerate needs to be an integral fraction -->
  <arg name="FPS" default="60/1"/>
  <!-- Available resolution should be 3448x808, 1748x408,1744x408,898x200,896x200 -->
  <arg name="width" default="3448"/>
  <arg name="height" default="808"/>
  <arg name="PUBLISH_FRAME" default="false"/>

  <!-- Split image to left and right image -->
  <node pkg="nodelet" type="nodelet" name="split_right" args="load image_proc/crop_decimate $(arg manager)">
    <param name="camera_info_url" value="$(arg camera_info_url_right)" />
    <param name="queue_size" type="int" value="10" />
    <param name="x_offset" type="int" value="48" />
    <param name="y_offset" type="int" value="0" />
    <param name="width" type="int" value="1280" />
    <param name="height" type="int" value="800" />
    <param name="camera_name" value="right"/>
    <!-- remap output to right image -->
    <remap from="camera_out/image_raw" to="/stereo/right/image_raw" />
    <!-- Dont use original camera info -->
    <remap from="/stereo/right/camera_info" to="/null/right/camera_info" />
  </node>
  <node pkg="nodelet" type="nodelet" name="split_left" args="load image_proc/crop_decimate $(arg manager)">
    <param name="camera_info_url" value="$(arg camera_info_url_left)" />
    <param name="queue_size" type="int" value="10" />
    <param name="x_offset" type="int" value="1328" />
    <param name="y_offset" type="int" value="0" />
    <param name="width" type="int" value="1280" />
    <param name="height" type="int" value="800" />
    <param name="camera_name" value="left"/>
    <!-- remap output to left image -->
    <remap from="camera_out/image_raw" to="/stereo/left/image_raw" />
    <!-- Dont use original camera info -->
    <remap from="/stereo/left/camera_info" to="/null/left/camera_info" />
  </node>

  <group ns="stereo">
    <group ns="disparity">
      <param name="disparity_range" type="double" value="128"/>
      <param name="speckle_size" type="double" value="100"/>
    </group>
    <include file="$(find stereo_image_proc)/launch/stereo_image_proc.launch">
      <arg name="manager" value="/$(arg manager)" />
    </include>
  </group>

  <arg name="PUBLISH_TF" default="true" />
  <node name="camera_transform" pkg="tf" type="static_transform_publisher" args="$(arg parent_transform) $(arg parent_frame) /ps4eye_frame 10" if="$(arg PUBLISH_TF)"/>

  <node pkg="ps4eye" type="camera_info_publisher.py" name="camera_info_publisher" >
    <param name="left_file_name"  value="$(arg camera_info_file_left)"  />
    <param name="right_file_name" value="$(arg camera_info_file_right)" />
  </node>

  <node pkg="nodelet" type="nodelet" name="resized_points" args="load jsk_pcl/ResizePointsPublisher $(arg manager)">
    <remap from="~input" to="/stereo/points2" />
    <param name="~step_x" value="8"/>
    <param name="~step_y" value="8"/>
  </node>
</launch>
